Sources/ITwinMobile/ITMOIDCAuthorizationClient.swift

```diff
@@ -295,47 +351,40 @@ open class ITMOIDCAuthorizationClient: NSObject, ITMAuthorizationClient, OIDAuth
         }
     }
 
-    open func signIn(_ completion: @escaping ITMOIDCAuthorizationClientCallback) {
+    private func innerSignIn(_ completion: @escaping ITMOIDCAuthorizationClientCallback) async {
         do {
             try checkSettings()
         } catch {
             completion(error)
             return
         }
-        guard let authSettings = authSettings else {
-            completion(createError(reason: "ITMOIDCAuthorizationClient not initialized"))
-            return
-        }
-        guard let issuerUrl = URL(string: authSettings.issuerUrl) else {
-            completion(createError(reason: "AuthSettings provided issuer URL that is invalid"))
-            return
-        }
-        if serviceConfig == nil {
-            OIDAuthorizationService.discoverConfiguration(forIssuer: issuerUrl) { [self] serviceConfig, error in
-                if let error = error {
-                    ITMApplication.logger.log(.error, "Failed to discover issuer configuration from \(issuerUrl): Error \(error)")
+        do {
+            let serviceConfig = try await requireServiceConfig()
+            if authState == nil {
+                doAuthCodeExchange(serviceConfig: serviceConfig, clientID: authSettings.clientId, clientSecret: nil) { error in
                     completion(error)
-                    return
                 }
-                self.serviceConfig = serviceConfig
-                signIn(completion)
-            }
-            return
-        }
-        if authState == nil {
-            doAuthCodeExchange(serviceConfig: serviceConfig, clientID: authSettings.clientId, clientSecret: nil) { error in
-                completion(error)
-            }
-        } else {
-            refreshAccessToken() { [self] error in
-                if error == nil {
-                    completion(error)
-                } else {
-                    // Refresh failed; sign out and try again from scratch.
-                    innerSignOut()
-                    signIn(completion)
+            } else {
+                refreshAccessToken() { [self] error in
+                    if error == nil {
+                        completion(error)
```

Comment:
I think this would be clearer if we passed nil to the completion rather than error (which is nil).